_posts/2019/2019-02-10-be-careful-assigning-to-has-one-relations.md

```diff
@@ -0,0 +1,132 @@
+=> #<Biography id: 2, person_id: 1, text: "Looks handsome" ...>
+Biography.count
+=> 2
+Biography.first
+=> #<Biography id: 1, person_id: nil, text: "Looks handsome" ...>
+```
+
+But I didn't call save?
+
+
+## Use…
+
+…a transaction to make sure the existing object is deleted if the new object is valid.
+
+```ruby
+class Biography < ApplicationRecord
+  belongs_to :person, optional: true
+
+  validates :text, presence: true
+end
+
+class Person < ApplicationRecord
+  has_one :biography
+
+  def replace_biography(attributes = nil)
+    Person.transaction do
+      biography&.destroy!
+      create_biography!(attributes)
+    end
+  rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotDestroyed
+    biography # returns invalid object
+  end
+end
+
+person = Person.create(name: "Andy Croll")
+=> #<Person id: 1, name: "Andy Croll" ...>
+person.create_biography(text: "Writes emails")
+=> #<Biography id: 1, person_id: 1, text: "Writes emails" ...>
+person.replace_biography(text: "Looks handsome")
+=> #<Biography id: 2, person_id: 1, text: Looks handsome ...>
+Biography.count
+=> 1
+```
+
+Boom.
+
+
+## But why?
+
+The first example demonstrates that the assignment (in this case via `#biography=`) to the `has_one` attribute removes the previously associated model and writes to the database.
+
+In Rails prior to 5.2.1 there is a bug that _even if_ the newly assigned model was invalid and would not save the existing record would still be changed.
+
+Normally when making assignments to the attributes of an Active Record model, via a `#whatever=` method, the changes are not persisted to the database until you call `#save`. In this association case, however, the change to the previously assigned model is written _at the point of assignment_.
+
+The method I’ve created in the second solution uses the `!` variants of the `create` and `destroy` methods inside a transaction to raise exceptions and rollback both the unassignment of the existing model and assignment of the new one if either fails.
+
+This prevents subtle bugs and the unusual behaviour of implicit writing to the database without calling `#save`.
+
+
+### In Modern Rails (after 5.0)
+
+In Rails 5.0, if you attempt to set the `person_id` attribute to `nil`, as the assignation of the new model will do, an `ActiveRecord::RecordNotSaved` error is raised. This prevents the assignation even if the new record is valid.
+
+The model being removed from the association is no longer valid because `belongs_to` relationships now have an [implied presence validation](https://guides.rubyonrails.org/5_0_release_notes.html#active-record-notable-changes). The change in Rails’ behaviour happened in [this pull request](https://github.com/rails/rails/pull/18937).
+
+I used the `optional: true` argument, in my example, to make the `Person` class behave more like pre-5.0 Rails.
+
+The solution I’ve proposed would also work in this more strict `belongs_to` relationship.
```

Comment:
Maybe you want to say it also works in newer versions of Rails? That's what you're getting at, right? But also — at this point I'm getting confused: with the stricter version, do you still need the fix?